### ISSUE TITLE
iOS Safari Advanced Tracking and Fingerprinting Protection breaks 3d features

### ISSUE THREAD

**Issue by az0uz** [mapbox/mapbox-gl-js#12898]
**mapbox-gl-js version**: v2.14.1

**browser**: iPhone Xs - Safari Mobile - iOS 17.0 

### Steps to Trigger Behavior

A. Check your 'Advanced Tracking and Fingerprinting Protection' settings, by default it might be only enabled for private browsing, but can be enabled for all browsing.

  1. Open up the Settings app.
  2. Scroll down to Safari.
  3. Tap on Advanced.
  4. Tap on Advanced Tracking and Fingerprinting Protection.
  5. Enable for "All Browsing." or keep it on "Private Browsing."

B. Open a mapbox demo with 3D terrain in safari

  1. Open Safari in either Private or Normal (depending on your 'Advanced Tracking and Fingerprinting Protection' settings)
  2. Navigate to a 3d enabled mapbox map, example: https://docs.mapbox.com/mapbox-gl-js/example/add-terrain/

### Link to Demonstration

Any map with 3d terrain, example: https://docs.mapbox.com/mapbox-gl-js/example/add-terrain/

### Expected Behavior
<img src="https://github.com/mapbox/mapbox-gl-js/assets/9094752/793f45c4-3ad0-4b6b-9fa5-8eace4eba18a" width="240px" />

### Actual Behavior
<img src="https://github.com/mapbox/mapbox-gl-js/assets/9094752/4399dbfe-10ee-47ad-99b0-d632b0d866d3" width="240px" />

**Comment by stepankuzmin** [comment]
I can confirm the issue. In the Private Browsing mode, Safari v17 on iOS blocks requests to the https://events.mapbox.com/events/v2. However, the `mapbox.mapbox-terrain-dem` source tiles are not blocked. Safari TP v17 on macOS seems to be working fine.

**Comment by mourner** [comment]
@stepankuzmin blocking events wouldn't cause the visual artifacts like on the screenshot; there must be something deeper happening in GL code than just network request blocking. Let's investigate

**Comment by az0uz** [comment]
The same issue in brave browser using `Agressively block tracker and Ads` mode happen, not as much though as you need to zoom-in more , which might be related.

See example this example when zooming-in on 3D terrain:

<img width="300" alt="screenshot of issue in rendering output with brave agressive tracker blocker" src="https://github.com/mapbox/mapbox-gl-js/assets/9094752/270b4f26-f6eb-4ec7-9b3d-b3798adbbf40">


If it can help, I think this is related to  the **farbling** feature in brave, which was also implemented in the latest safari
see related resources for brave
https://brave.com/privacy-updates/4-fingerprinting-defenses-2.0/#farbling
https://github.com/brave/brave-browser/issues/17850

TLDR, noise is applied when reading pixel values from the GPU

**Comment by tryonelove** [comment]
The same issue occurs with Helium when using its 'Helium Noise' flag
